### PR TITLE
Make SHA256 tests more explicit

### PR DESCRIPTION
**Pull request by smanilov**
The tests now reuse the code for the test logic. They differ only in the input data and the expected messages.

The inputs for the tests are now explicitly specified as bytes in hex, as well as the expected output. This surfaces that the last byte of the succeeding test is 0x00, which might be not have been the intention, reading the original code.

Randomness is removed from the failing test, as it is an unnecessary complication and reduces reproducibility of the test.

A new test is added that makes sure the SHA256 implementation corresponds to that of coreutils-9.3 for one real ASCII string.

### CODE REVIEW

**Review by Dimo99 — DISMISSED** [review]

File: beacon-light-client/plonky2/circuits/src/sha256.rs [hunk 1]
```diff
@@ -399,49 +399,29 @@ pub fn make_circuits<F: RichField + Extendable<D>, const D: usize>(
...
-        let mut msg = vec![0; 128 as usize];
-        for i in 0..127 {
-            msg[i] = i as u8;
-        }
+    use super::super::sha256::{array_to_bits, make_circuits};
 
-        let msg_bits = array_to_bits(&msg);
-        let len = msg.len() * 8;
+    fn perform_sha256_test(message: &[u8], expected: &[u8]) {
+        let message_bits = array_to_bits(message);
```
Comment (Xearty) [inline beacon-light-client/plonky2/circuits/src/sha256.rs:410]: While you are at it, could you rename `array_to_bits` to `bytes_to_bits`. I have also written a sha256 circuit function with alternative api so we will revisit these tests but you could rename it now so we don't forget later.

File: beacon-light-client/plonky2/circuits/src/sha256.rs [hunk 2]
```diff
@@ -451,44 +431,77 @@ mod tests {
         let data = builder.build::<C>();
         let proof = data.prove(pw).unwrap();
 
-        data.verify(proof)
+        data.verify(proof).expect("");
```
Comment (Xearty) [inline beacon-light-client/plonky2/circuits/src/sha256.rs:434]: Just `unwrap()` here

File: beacon-light-client/plonky2/circuits/src/sha256.rs [hunk 3]
```diff
@@ -451,44 +431,77 @@ mod tests {
...
+        let mut expected: [u8; 32] = [
+            0x3c, 0x5e, 0xa3, 0xe9, 0x28, 0xf9, 0x36, 0x43, 0xe6, 0xf5, 0x70, 0x56, 0x99, 0x16,
+            0xba, 0x9f, 0x71, 0x56, 0x0e, 0x70, 0x84, 0x41, 0x38, 0x55, 0xb4, 0xa0, 0x2d, 0xd7,
+            0x5d, 0x38, 0x3e, 0x02,
+        ];
+
+        // Modify some byte to lead to a wrong result.
+        expected[10] = 0xff;
+
+        perform_sha256_test(&MESSAGE, &expected);
```
Comment (Xearty) [inline beacon-light-client/plonky2/circuits/src/sha256.rs:505]: This macro can be used in the tests. It creates a byte vec from hex string
```rs
pub macro bytes($hex_literal:expr) {{
    let hex_string = $hex_literal;
    let stripped = if let Some(stripped) = hex_string.strip_prefix("0x") {
        stripped
    } else {
        &hex_string
    };
    hex::decode(stripped)
        .expect("Invalid hex string")
        .try_into()
        .expect(&format!(
            "Wrong byte length {} for hex string {}",
            stripped.len(),
            hex_string
        ))
}}
```

Comment (smanilov) [reply beacon-light-client/plonky2/circuits/src/sha256.rs:505]: Tried to do it as a pair with the reviewer and it proved more involved than intended.

Will not do.

**Review by Xearty — APPROVED** [review]

**Review by Dimo99 — APPROVED** [review]

**Comment by smanilov** [comment]
Could you please merge this, since I don't have write access. Thanks.

**Pull request merged**